File: search/array-based-search/binary-search-on-answer/python/ship_packages.py
def can_ship(weights, capacity, max_days):
    """
    Check if all packages can be shipped within 'max_days' using given 'capacity'.
    
    Simulates loading in order, starting new days when capacity is exceeded.
    
    Args:
        weights (list): List of package weights (in order)
        capacity (int): Ship capacity per day
        max_days (int): Maximum allowed days
    
    Returns:
        bool: True if shipping within max_days is possible
    """
    days_needed = 1
    current_load = 0
    
    for w in weights:
        if current_load + w <= capacity:
            current_load += w
        else:
            days_needed += 1
            current_load = w
            if days_needed > max_days:
                return False  # Early exit
    
    return days_needed <= max_days


def min_ship_capacity(weights, D):
    """
    Finds the minimum ship capacity required to ship all packages in exactly D days.
    
    Uses binary search on the answer (capacity) between:
      low = max(weights)  -> at least carry heaviest package
      high = sum(weights) -> ship all in one day
    
    Args:
        weights (list): Package weights in delivery order
        D (int): Number of days to ship all packages
    
    Returns:
        int: Minimum capacity required
    """
    if not weights or D <= 0:
        raise ValueError("Invalid input: weights must be non-empty and D > 0")
    if D > len(weights):
        raise ValueError("D cannot be greater than number of packages")

    low = max(weights)
    high = sum(weights)
    result = high

    print("\n🔍 Binary Search Trace:")
    print(f"{'Iter':<4} {'Low':<4} {'High':<4} {'Mid':<4} {'Feasible?':<10} Action")
    print("-" * 50)

    iter_count = 1
    while low <= high:
        mid = (low + high) // 2
        feasible = can_ship(weights, mid, D)
        mark = "✅" if feasible else "❌"
        status = "Yes" if feasible else "No"

        if feasible:
            txt = f"{iter_count:<4} {low:<4} {high:<4}"
            result = mid
            action = f"Try smaller → high = {mid - 1}"
            high = mid - 1
        else:
            txt = f"{iter_count:<4} {low:<4} {high:<4}"
            action = f"Need larger → low = {mid + 1}"
            low = mid + 1

        print(f"{txt} {mid:<4} "
              f"{mark + ' ' + status:<10} {action}")
        
        iter_count += 1

    return result


def simulate_shipping(weights, capacity, D):
    """
    Simulate and print how packages are shipped with given capacity.
    
    Args:
        weights (list): Package weights
        capacity (int): Ship capacity
        D (int): Allowed days
    """
    print(f"\n📦 Shipping Simulation (Capacity = {capacity}, Max Days = {D}):")
    day = 1
    current_load = 0
    daily_packages = [[]]

    for w in weights:
        if current_load + w <= capacity:
            current_load += w
            daily_packages[-1].append(w)
        else:
            day += 1
            current_load = w
            daily_packages.append([w])

    total_days = len(daily_packages)
    valid = total_days <= D

    for i, pkgs in enumerate(daily_packages, 1):
        load = sum(pkgs)
        print(f"  Day {i}: [{', '.join(map(str, pkgs))}] → Total = {load} {'✅' if load <= capacity else '❌'}")

    print(f"🎯 Total Days Used: {total_days} {'✅' if valid else '❌ Exceeds Limit'}")


# Example usage and demonstration
if __name__ == "__main__":
    print("🔍 Binary Search on Answer: Ship Packages Within D Days\n")
    
    weights = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    D = 5
    
    print(f"Weights: {weights}")
    print(f"Days Available: {D}")
    print(f"Min Weight: {min(weights)} | Max Weight: {max(weights)} | Total: {sum(weights)}")

    min_capacity = min_ship_capacity(weights, D)
    print(f"\n✅ Minimum Required Capacity: {min_capacity}")

    simulate_shipping(weights, min_capacity, D)
    print(f"\n🎯 Final Answer: The smallest capacity that works is {min_capacity}.")

File: search/array-based-search/binary-search-on-answer/javascript/ship_packages.js
/**
 * Checks if all packages can be shipped within 'maxDays' using given 'capacity'.
 * Simulates loading in order, starting new days when capacity is exceeded.
 *
 * @param {number[]} weights - Array of package weights (in order)
 * @param {number} capacity - Ship capacity per day
 * @param {number} maxDays - Maximum allowed days
 * @returns {boolean} - True if shipping within maxDays is possible
 */
function canShip(weights, capacity, maxDays) {
    let daysNeeded = 1;
    let currentLoad = 0;

    for (const w of weights) {
        if (currentLoad + w <= capacity) {
            currentLoad += w;
        } else {
            daysNeeded++;
            currentLoad = w;
            // Early exit if already over limit
            if (daysNeeded > maxDays) return false;
        }
    }

    return daysNeeded <= maxDays;
}

/**
 * Finds the minimum ship capacity required to ship all packages in exactly D days.
 * Uses binary search on the answer (capacity) between:
 *   low = max(weights)  -> at least carry heaviest package
 *   high = sum(weights) -> ship all in one day
 *
 * @param {number[]} weights - List of package weights (in order)
 * @param {number} D - Number of days to ship all packages
 * @returns {number} - Minimum capacity required
 */
function minShipCapacity(weights, D) {
    if (!weights || weights.length === 0 || D <= 0) {
        throw new Error("Invalid input: weights must be non-empty and D > 0");
    }
    if (D > weights.length) {
        throw new Error("D cannot be greater than number of packages");
    }

    const low = Math.max(...weights);     // Must carry heaviest single package
    const high = weights.reduce((a, b) => a + b, 0); // Can ship all in one day

    let result = high;

    console.log(`\n🔍 Binary Search Trace:`);
    console.log(`${'Iter'.padEnd(4)} ${'Low'.padEnd(4)} ${'High'.padEnd(4)} ${'Mid'.padEnd(4)} ${'Feasible?'.padEnd(10)} Action`);
    console.log("-".repeat(50));

    let left = low;
    let right = high;
    let iter = 1;

    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        const feasible = canShip(weights, mid, D);
        const mark = feasible ? "✅" : "❌";
        const status = feasible ? "Yes" : "No";
        let action;
        let txt;

        if (feasible) {
            txt =`${iter.toString().padEnd(4)} ${left.toString().padEnd(4)} ${right.toString().padEnd(4)}`
            result = mid;
            right = mid - 1;
            action = `Try smaller → high = ${right}`;
           } else {
            txt =`${iter.toString().padEnd(4)} ${left.toString().padEnd(4)} ${right.toString().padEnd(4)} `
            left = mid + 1;
            action = `Need larger → low = ${left}`;
        }

        console.log(`${txt}${mid.toString().padEnd(4)} ${mark + ' ' + status.padEnd(8)} ${action}`);
        iter++;
    }

    return result;
}

/**
 * Simulate and print how packages are shipped with given capacity.
 *
 * @param {number[]} weights - Package weights
 * @param {number} capacity - Ship capacity
 * @param {number} D - Allowed days
 */
function simulateShipping(weights, capacity, D) {
    console.log(`\n📦 Shipping Simulation (Capacity = ${capacity}, Max Days = ${D}):`);
    let day = 1;
    let currentLoad = 0;
    const dailyPackages = [[]];

    for (const w of weights) {
        if (currentLoad + w <= capacity) {
            currentLoad += w;
            dailyPackages[dailyPackages.length - 1].push(w);
        } else {
            day++;
            currentLoad = w;
            dailyPackages.push([w]);
        }
    }

    const totalDays = dailyPackages.length;
    const valid = totalDays <= D;

    dailyPackages.forEach((pkgs, i) => {
        const load = pkgs.reduce((a, b) => a + b, 0);
        console.log(`  Day ${i + 1}: [${pkgs.join(", ")}] → Total = ${load} ${load <= capacity ? '✅' : '❌'}`);
    });

    console.log(`🎯 Total Days Used: ${totalDays} ${valid ? '✅' : '❌ Exceeds Limit'}`);
}

// Example usage and demonstration
(() => {
    console.log("🔍 Binary Search on Answer: Ship Packages Within D Days\n");

    const weights = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10];
    const D = 5;

    console.log(`Weights: [${weights.join(", ")}]`);
    console.log(`Days Available: ${D}`);
    console.log(`Min Weight: ${Math.min(...weights)} | Max Weight: ${Math.max(...weights)} | Total: ${weights.reduce((a, b) => a + b, 0)}`);

    // Find minimum capacity
    const minCapacity = minShipCapacity(weights, D);
    console.log(`\n✅ Minimum Required Capacity: ${minCapacity}`);

    // Verify with simulation
    simulateShipping(weights, minCapacity, D);

    // Final confirmation
    console.log(`\n🎯 Final Answer: The smallest capacity that works is ${minCapacity}.`);
})();

module.exports ={minShipCapacity,canShip}